crates/turbopack-convert-trace/src/main.rs

```diff
@@ -236,34 +236,73 @@ fn main() {
+                        thread_id: parent_thread_id,
+                        ..
+                    }) = parent.self_start
+                    {
+                        if parent_thread_id == thread_id {
+                            let ts_start = *parent_ts;
+                            *parent_ts = ts;
+                            add_self_time(
+                                ts_start,
+                                ts,
+                                internal_id,
+                                parent,
+                                &mut all_self_times,
+                                &mut name_self_times,
+                            );
+                            in_parent = parent_id;
+                            break;
+                        }
+                    }
+                    if parent_id == 0 {
+                        break;
+                    }
+                    parent_id = parent.parent;
+                }
                 let span = &mut spans[internal_id];
-                span.self_start = Some(SelfTimeStarted { ts });
+                span.self_start = Some(SelfTimeStarted {
+                    ts,
+                    thread_id,
+                    in_parent,
+                });
             }
             TraceRow::Exit { ts, id } => {
                 let internal_id = ensure_span(&mut active_ids, &mut spans, id);
                 let span = &mut spans[internal_id];
-                if let Some(SelfTimeStarted { ts: ts_start }) = span.self_start {
-                    let (start, end) = if ts_start > ts {
-                        (ts, ts_start)
-                    } else {
-                        (ts_start, ts)
-                    };
-                    let duration = end.saturating_sub(start);
-                    span.items.push(SpanItem::SelfTime { start, duration });
-                    if duration > 0 {
-                        all_self_times.push(Element {
-                            range: start..end,
-                            value: internal_id,
-                        });
+                if let Some(SelfTimeStarted {
+                    ts: ts_start,
+                    in_parent,
+                    ..
+                }) = span.self_start.take()
+                {
+                    add_self_time(
+                        ts_start,
+                        ts,
+                        internal_id,
+                        span,
+                        &mut all_self_times,
+                        &mut name_self_times,
+                    );
+                    if in_parent > 0 {
+                        let parent_id = span.parent;
+                        let span = &mut spans[parent_id];
+                        if let Some(SelfTimeStarted {
+                            ts: ref mut parent_ts,
+                            ..
+                        }) = span.self_start
+                        {
+                            *parent_ts = max(ts, ts_start);
```

Comment:
I don't understand this. We're setting the parent's starting timestamp to be later than originally reported?

Reply:
Example:

Enter span A
1s
Enter span B
1s
Exit span B
1s
Exit span A


Before this change it collects 3s self time for span A and 1s self time of span B. But that is not correct. We need to subtract the 1s spend in span B from span A when they execute on the same thread.

So when we enter span B and there is already a parent span active on the same thread, we "exit" the parent span (A). Once we exit the inner span (B) again, we "enter" the parent span again. This "enter" is handled on that line. We enter the parent by setting the self start time to the current time.